### PR TITLE
use GHA concurrency

### PR DESCRIPTION
**Pull request by joshgubler**

### CODE REVIEW

**Comment by joshgubler** [comment]
Well it didn't break. Probably need to push a few commits in quick succession to test if it actually works.

File: .github/workflows/ci.yml
Comment (GaryGSC) [inline .github/workflows/ci.yml:10]: I thought we were thinking of putting this on just the Terraform Plan job.

**Review by GaryGSC — APPROVED** [review]
I'm fine if we merge this into `dev` so that we can test things running at the same time as a deployment. After we've tested it, I'd still like to narrow the scope of our concurrency constraints in CI.

**Review by jvisker — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (joshgubler) [reply .github/workflows/ci.yml:10]: I was originally planning on that, but this note scared me

> Note: When concurrency is specified at the job level, order is not guaranteed for jobs or runs that queue within 5 minutes of each other.

Do we want to ensure that the last commit to a PR actually gets planned?

**Comment by GaryGSC** [comment]
Re-running CI on #172 might be a good test case.

**Comment by joshgubler** [comment]
> Re-running CI on #172 might be a good test case.

That run didn't have the GHA concurrency set, so I expect it would run freely.

File: .github/workflows/ci.yml
Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: I assumed that was to account for the way jobs don't always start right away (and so therefore we have a race condition). I feel like I'd prefer most of our CI to run rather than having it all blocked on a deployment.

**Comment by GaryGSC** [comment]
🤔 That makes sense.

File: .github/workflows/ci.yml
Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: Worst case scenario, "previous" jobs that were already blocked get cancelled and we re-run CI. That doesn't sound too bad.

Actually, I guess the real worst-case scenario is if we have a deployment blocked on CI and we update our PR that had CI running, we cancel the deployment. We'd then have to re-run the deployment. Not a problem, but still annoying.

Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: Hopefully, the Slack notification would make it obvious to a developer that their queued deployment was cancelled.

Comment (joshgubler) [reply .github/workflows/ci.yml:10]: I think worst case is that somebody pushes several commits to a PR in succession, the last commit contains a mistake that breaks their infrastructure, but the CI plan doesn't catch the problem because the wrong run survives. So the broken infrastructure gets deployed.

I don't know how likely this is to happen. But 5 minutes is a long time when you're pushing small changes fast.

Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: > When a concurrent job or workflow is queued, if another job or workflow using the same concurrency group in the repository is in progress, the queued job or workflow will be pending. Any previously pending job or workflow in the concurrency group will be canceled.

([source](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idconcurrency))

> the CI plan doesn't catch the problem

I'm not sure I understand the concern here. If CI is cancelled, that shows up as "failing" CI. If CI is pending, it won't show up as green until it's finished running.

Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: Check out the `update prd files` commit [here](https://github.com/byu-oit/casfsy/commits/main) on April 20th. You can see it marked as ❌ because the deployment was cancelled (without being re-run).

Comment (GaryGSC) [reply .github/workflows/ci.yml:10]: If CI is cancelled, I'd expect developers to re-run it on their PR to get a ✔ before merging the PR.

**Comment by GaryGSC** [comment]
@joshgubler - Can we pick this PR back up?

I'm not sure where I stand on the interoperability between the `CI / Terraform Plan` job and the `Deploy` workflow, but even just using `concurrency` on the `Deploy` workflow is a strict improvement over what we already have.

**Comment by jvisker** [comment]
Would this change mean that a deploy would block a CI from running?

**Comment by GaryGSC** [comment]
I'd also be fine with us taking off the `concurrency` bit from the Terraform Plan job and only keeping the `deploy.yml` changes. That was the strict improvement case I was mentioning above, since `concurrency` should work better than the `byu-oit/github-action-disallow-concurrent-runs@v2` action.

**Review by blakemorgan — COMMENTED** [review]
I currently use `concurrency` on the deploy workflows and then separated my infrastructure CI into a different workflow from my application CI. While the infrastructure CI and the deployment _could_ still run at the same (causing a failed plan), I've noticed it happening a lot less since I don't check the infrastructure every time I change the application. That may be a possible compromise we can use here.

**Comment by GaryGSC** [comment]
@byu-oit/devops-tooling - Can we close this now that we've merged #222 and #327?

A deployment obtaining the TF state lock might still cause CI to fail, but CI shouldn't conflict with any other workflows and deployments shouldn't conflict with other deployments.

**Pull request closed without merging**
